### PR TITLE
Chapter 9: Advanced Login

### PR DESCRIPTION
**Pull request by puteriffah**

### CODE REVIEW

**Comment by puteriffah** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -106,3 +106,17 @@ input.btn.btn-primary {
     color: $state-danger-text;
   }
 }
+
+.checkbox {
+  margin-bottom: 10px;
+  margin-top: 10px;
+  span {
+  font-weight: normal;
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: image

File: app/assets/stylesheets/sessions.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/sessions.scss]: image

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,3 +1,19 @@
 class ApplicationController < ActionController::Base
   include SessionsHelper
+
+  before_action :set_locale
+
+  private
+  def set_locale
```
Comment (saobv-2223) [inline app/controllers/application_controller.rb]:
```suggestion
  private
  
  def set_locale
```

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -1,11 +1,12 @@
...
 
   def create
-    user = User.find_by email: params[:session][:email].downcase
-    if user&.authenticate params[:session][:password]
-      log_in user
-      redirect_to user_path(id: current_user.id)
+    @user = User.find_by email: params[:session][:email].downcase
+    if @user&.authenticate params[:session][:password]
+      log_in @user
+      params[:session][:remember_me] == "1" ? remember(@user) : forget(@user)
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]: using gem config to manage number 1

File: app/models/user.rb [hunk 5]
```diff
@@ -16,9 +17,38 @@ class User < ApplicationRecord
...
+  def forget
+    update_attribute :remember_digest, nil
+  end
+
   private
 
   def downcase_email
     email.downcase!
   end
+  before_save { email.downcase! }
```
Comment (saobv-2223) [inline app/models/user.rb]: move above private

**Comment by puteriffah** [comment]
ready

File: app/assets/stylesheets/sessions.scss (hunk 2 above)
Comment (ngocvt-0484) [reply app/assets/stylesheets/sessions.scss]: remove this file because you have not used it

File: app/models/user.rb [hunk 6]
```diff
@@ -16,6 +17,36 @@ class User < ApplicationRecord
...
+
+  def authenticated? remember_token
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+
+  def forget
+    update_attribute :remember_digest, nil
+  end
+
+  before_save { email.downcase! }
```
Comment (ngocvt-0484) [inline app/models/user.rb]: line 18 you have downcase function
so dont need this line
please remove it

**Comment by puteriffah** [comment]
ready

File: app/models/user.rb [hunk 7]
```diff
@@ -16,6 +17,34 @@ class User < ApplicationRecord
 
   before_save :downcase_email
 
+  class << self
+    def new_token
+    SecureRandom.urlsafe_base64
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
      SecureRandom.urlsafe_base64
```

**Comment by puteriffah** [comment]
ready

**Comment by manhnt-0960** [comment]
merged

**Pull request merged**